sidecar/src/agentic/tool/mcp/integration_tool.rs

```diff
@@ -0,0 +1,195 @@
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::{collections::HashMap, sync::Arc};
+
+use crate::agentic::tool::{
+    errors::ToolError,
+    input::ToolInput,
+    output::ToolOutput,
+    r#type::{Tool, ToolRewardScale},
+};
+use mcp_client_rs::client::Client;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ToolDescriptor {
+    pub name: String,
+    #[serde(default)]
+    pub description: Option<String>,
+    #[serde(default)]
+    pub schema: Option<Value>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ServerTools {
+    pub server_name: String,
+    pub tools: Vec<ToolDescriptor>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ToolListResponse {
+    pub servers: Vec<ServerTools>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ToolCallResponse {
+    pub result: Value,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "type", content = "data")]
+pub enum MCPIntegrationToolResponse {
+    ToolList(ToolListResponse),
+    ToolCall(ToolCallResponse),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct MCPIntegrationToolQuery {
+    pub action: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub server_name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub tool_name: Option<String>,
+    #[serde(default)]
+    pub arguments: Value,
+}
+
+impl MCPIntegrationToolQuery {
+    pub fn to_json() -> Value {
+        serde_json::json!({
+            "action": "list | call",
+            "server_name": "string (required if action=call)",
+            "tool_name": "string (required if action=call)",
+            "arguments": {}
+        })
+    }
+}
+
+pub struct MCPIntegrationToolBroker {
+    servers: HashMap<String, Arc<Client>>,
+}
+
+impl MCPIntegrationToolBroker {
+    pub fn new(servers: HashMap<String, Arc<Client>>) -> Self {
+        Self { servers }
+    }
+
+    async fn list_all_tools(&self) -> Result<MCPIntegrationToolResponse, ToolError> {
+        let mut server_list = Vec::new();
+        for (server_name, client) in &self.servers {
+            let tools_value = client.list_tools().await.map_err(|e| {
+                ToolError::InvocationError(format!(
+                    "Failed to list tools from '{}': {}",
+                    server_name, e
+                ))
+            })?;
+
+            let tools_array = tools_value
+                .get("tools")
+                .and_then(Value::as_array)
+                .ok_or_else(|| {
+                    ToolError::InvocationError("Missing 'tools' array in response".to_string())
+                })?;
+
+            let mut tool_descriptors = Vec::new();
+            for t in tools_array {
```

Comment:
proper variable names please, I dont enjoy second guessing single variable names and looking for them in the codebase.. there is no need to be stingy with the variable name